src/transaction/log_manager.c

```diff
@@ -3824,7 +3836,8 @@ log_sysop_commit_internal (THREAD_ENTRY * thread_p, LOG_REC_SYSOP_END * log_reco
        * we don't actually allow empty logical system operation because it might hide a logic flaw. however, there are
        * unusual cases when a logical operation does not really require logging (see RVPGBUF_FLUSH_PAGE). if you create
        * such a case, you should add a dummy log record to trick this assert. */
-      assert (!LSA_ISNULL (&tdes->tail_lsa) && LSA_GT (&tdes->tail_lsa, LOG_TDES_LAST_SYSOP_PARENT_LSA (tdes)));
+      assert (log_No_logging || (!LSA_ISNULL (&tdes->tail_lsa)
```

Comment:
I doubt it is correct and safe.. It must be the last option to drop or change the existing assertion. The assert makes much sense to me. In this `else`, a logical undo log record will be appended below, but if there is no redo log records have been generated for the logical operation, it means that there is an logical UNDO for no physical REDO modifications. The UNDO assumes that there are proper physical changes.

Analyzing related-code, most of time, it would be not a problem because during `log_No_logging`, if a crash occurs, it just stops the entire RECOVERY or ROLLBACK with a fatal error. However, it's wrong logically, so it has a potential error. An example would be the `log_sysop_abort`, which seems try to undo even during `log_No_logging` when a sysop encounters an internal error.

The way I see it, it does more make sense to change the `if` state for no log record above
` if ((LSA_ISNULL (&tdes->tail_lsa) || LSA_LE (&tdes->tail_lsa, LOG_TDES_LAST_SYSOP_PARENT_LSA (tdes)))
      && log_record->type == LOG_SYSOP_END_COMMIT)`
->
` if ((LSA_ISNULL (&tdes->tail_lsa) || LSA_LE (&tdes->tail_lsa, LOG_TDES_LAST_SYSOP_PARENT_LSA (tdes)))
      && (log_record->type == LOG_SYSOP_END_COMMIT || log_No_logging))`

However, it needs some additional tests for the change. Please share your opinion on it.

Reply:
I don't know much about recovery and log management.
All I know is that this function is supposed to not return an error and assert if something goes wrong.
Therefore, assert is only made in case a critical error occurs.
Your opinion is to use an if statement instead of assert, but ultimately else should be handled as assert(false).

Reply:
> Your opinion is to use an if statement instead of assert, but ultimately else should be handled as assert(false).

I don't understand what you mean. What I meant is even if `log_No_logging` is true, the existing assert should be met. The issue case will be go through the `if` case.

Reply:
I misunderstood your comment.
I now understand that it is not an assert but a higher level if statement. I aggree with you.

Reply:
This assert should be reverted as well.

Reply:
I modified it as your comment.